File: apps/desktop/src-tauri/src/sys/commands/governance.rs
use crate::data::db::Database;
use crate::sys::error::Result;
use crate::sys::security::audit_logger::AuditFilters;
use crate::sys::security::{
    create_tool_execution_event, create_workflow_execution_event, ApprovalAction, ApprovalDecision,
    ApprovalRequest, ApprovalStatistics, ApprovalWorkflow, AuditEvent, AuditIntegrityReport,
    AuditStatus, EnhancedAuditLogger,
};
use tauri::State;

#[tauri::command]
pub async fn get_audit_events(
    filters: AuditFilters,
    db: State<'_, Database>,
) -> Result<Vec<AuditEvent>> {
    let conn = db.get_connection();
    let logger = EnhancedAuditLogger::new(conn)?;

    logger.get_events(filters)
}

#[tauri::command]
pub async fn verify_audit_event(event_id: String, db: State<'_, Database>) -> Result<bool> {
    let conn = db.get_connection();
    let logger = EnhancedAuditLogger::new(conn)?;

    logger.verify_event(&event_id)
}

#[tauri::command]
pub async fn verify_audit_integrity(db: State<'_, Database>) -> Result<AuditIntegrityReport> {
    let conn = db.get_connection();
    let logger = EnhancedAuditLogger::new(conn)?;

    logger.verify_all_events()
}

#[tauri::command]
pub async fn log_tool_execution(
    user_id: Option<String>,
    team_id: Option<String>,
    tool_name: String,
    success: bool,
    metadata: Option<serde_json::Value>,
    db: State<'_, Database>,
) -> Result<()> {
    let conn = db.get_connection();
    let logger = EnhancedAuditLogger::new(conn)?;

    let event = create_tool_execution_event(user_id, team_id, tool_name, success, metadata);

    logger.log(event)
}

#[tauri::command]
pub async fn log_workflow_execution(
    user_id: Option<String>,
    team_id: Option<String>,
    workflow_id: String,
    status: String,
    metadata: Option<serde_json::Value>,
    db: State<'_, Database>,
) -> Result<()> {
    let conn = db.get_connection();
    let logger = EnhancedAuditLogger::new(conn)?;

    let audit_status = match status.as_str() {
        "success" => AuditStatus::Success,
        "failure" => AuditStatus::Failure,
        "blocked" => AuditStatus::Blocked,
        "pending" => AuditStatus::Pending,
        _ => AuditStatus::Pending,
    };

    let event =
        create_workflow_execution_event(user_id, team_id, workflow_id, audit_status, metadata);

    logger.log(event)
}

#[tauri::command]
pub async fn create_approval_request(
    app_handle: tauri::AppHandle,
    requester_id: String,
    team_id: Option<String>,
    action: ApprovalAction,
    risk_level: String,
    justification: Option<String>,
    timeout_minutes: i64,
    db: State<'_, Database>,
) -> Result<String> {
    let conn = db.get_connection();
    let workflow = ApprovalWorkflow::new(conn);

    let risk = match risk_level.as_str() {
        "low" => crate::sys::security::ApprovalRiskLevel::Low,
        "medium" => crate::sys::security::ApprovalRiskLevel::Medium,
        "high" => crate::sys::security::ApprovalRiskLevel::High,
        "critical" => crate::sys::security::ApprovalRiskLevel::Critical,
        _ => crate::sys::security::ApprovalRiskLevel::Medium,
    };

    let request_id = workflow.create_approval_request(
        requester_id,
        team_id,
        action.clone(),
        risk,
        justification.clone(),
        timeout_minutes,
    )?;

    let created_at = chrono::Utc::now();
    let approval_payload = crate::ui::events::ApprovalRequestPayload {
        id: request_id.clone(),
        request_type: action.action_type.clone(),
        description: format!(
            "{} on {}",
            action.action_type,
            action
                .resource_id
                .as_ref()
                .unwrap_or(&"unknown resource".to_string())
        ),
        impact: justification.clone(),
        risk_level: risk_level.to_lowercase(),
        status: "pending".to_string(),
        created_at: created_at.to_rfc3339(),
        approved_at: None,
        rejected_at: None,
        rejection_reason: None,
        timeout_seconds: Some(timeout_minutes * 60),
        details: Some(serde_json::to_value(&action).unwrap_or(serde_json::json!({}))),
    };

    crate::ui::events::emit_approval_request(&app_handle, approval_payload);

    Ok(request_id)
}

#[tauri::command]
pub async fn get_pending_approvals(
    team_id: Option<String>,
    db: State<'_, Database>,
) -> Result<Vec<ApprovalRequest>> {
    let conn = db.get_connection();
    let workflow = ApprovalWorkflow::new(conn);

    workflow.get_pending_approvals(team_id)
}

#[tauri::command]
pub async fn get_approval_request(
    request_id: String,
    db: State<'_, Database>,
) -> Result<ApprovalRequest> {
    let conn = db.get_connection();
    let workflow = ApprovalWorkflow::new(conn);

    workflow.get_request(&request_id)
}

#[tauri::command]
pub async fn approve_request(
    request_id: String,
    reviewer_id: String,
    reason: Option<String>,
    db: State<'_, Database>,
) -> Result<()> {
    let conn = db.get_connection();
    let workflow = ApprovalWorkflow::new(conn);

    let decision = ApprovalDecision::Approved { reason };

    workflow.approve_request(&request_id, &reviewer_id, decision)
}

#[tauri::command]
pub async fn reject_request(
    request_id: String,
    reviewer_id: String,
    reason: String,
    db: State<'_, Database>,
) -> Result<()> {
    let conn = db.get_connection();
    let workflow = ApprovalWorkflow::new(conn);

    let decision = ApprovalDecision::Rejected { reason };

    workflow.approve_request(&request_id, &reviewer_id, decision)
}

#[tauri::command]
pub async fn requires_approval(action: ApprovalAction, db: State<'_, Database>) -> Result<bool> {
    let conn = db.get_connection();
    let workflow = ApprovalWorkflow::new(conn);

    Ok(workflow.requires_approval(&action))
}

#[tauri::command]
pub async fn calculate_risk_level(
    action: ApprovalAction,
    db: State<'_, Database>,
) -> Result<String> {
    let conn = db.get_connection();
    let workflow = ApprovalWorkflow::new(conn);

    let risk = workflow.calculate_risk_level(&action);

    Ok(risk.as_str().to_string())
}

#[tauri::command]
pub async fn get_approval_statistics(
    team_id: Option<String>,
    db: State<'_, Database>,
) -> Result<ApprovalStatistics> {
    let conn = db.get_connection();
    let workflow = ApprovalWorkflow::new(conn);

    workflow.get_statistics(team_id)
}

#[tauri::command]
pub async fn expire_timed_out_requests(db: State<'_, Database>) -> Result<usize> {
    let conn = db.get_connection();
    let workflow = ApprovalWorkflow::new(conn);

    workflow.expire_timed_out_requests()
}

File: packages/api/src/governance.ts
/**
 * Governance & Privacy API — typed wrappers for audit, approval, and privacy commands.
 */

import { command } from '@agiworkforce/runtime';

// ---- Types ----

export interface AuditFilters {
  userId?: string;
  teamId?: string;
  action?: string;
  startDate?: string;
  endDate?: string;
  limit?: number;
}
export interface AuditEvent {
  id: string;
  userId: string;
  action: string;
  details: unknown;
  timestamp: string;
  hash: string;
}
export interface AuditIntegrityReport {
  totalEvents: number;
  verifiedEvents: number;
  corruptedEvents: number;
}
export interface ApprovalAction {
  type: string;
  resourceId?: string;
  details?: unknown;
}
export interface ApprovalRequest {
  id: string;
  requesterId: string;
  action: ApprovalAction;
  riskLevel: string;
  status: string;
  justification?: string;
  createdAt: string;
}
export interface ApprovalStatistics {
  total: number;
  pending: number;
  approved: number;
  rejected: number;
  avgResponseTime: number;
}
export interface PrivacyPreferences {
  telemetryEnabled: boolean;
  crashReportingEnabled: boolean;
  aiModelSharingEnabled: boolean;
  analyticsEnabled: boolean;
  usageDataCollection: boolean;
}

// ---- Audit ----

export async function getAuditEvents(filters: AuditFilters): Promise<AuditEvent[]> {
  return command<AuditEvent[]>('get_audit_events', { filters });
}
export async function verifyAuditEvent(eventId: string): Promise<boolean> {
  return command<boolean>('verify_audit_event', { eventId });
}
export async function verifyAuditIntegrity(): Promise<AuditIntegrityReport> {
  return command<AuditIntegrityReport>('verify_audit_integrity');
}
export async function logToolExecution(
  toolName: string,
  success: boolean,
  userId?: string,
  teamId?: string,
  metadata?: unknown,
): Promise<void> {
  return command<void>('log_tool_execution', { userId, teamId, toolName, success, metadata });
}
export async function logWorkflowExecution(
  workflowId: string,
  status: string,
  userId?: string,
  teamId?: string,
  metadata?: unknown,
): Promise<void> {
  return command<void>('log_workflow_execution', { userId, teamId, workflowId, status, metadata });
}

// ---- Approvals ----

export async function createApprovalRequest(
  requesterId: string,
  action: ApprovalAction,
  riskLevel: string,
  timeoutMinutes: number,
  teamId?: string,
  justification?: string,
): Promise<string> {
  return command<string>('create_approval_request', {
    requesterId,
    teamId,
    action,
    riskLevel,
    justification,
    timeoutMinutes,
  });
}
export async function getPendingApprovals(teamId?: string): Promise<ApprovalRequest[]> {
  return command<ApprovalRequest[]>('get_pending_approvals', { teamId });
}
export async function getApprovalRequest(requestId: string): Promise<ApprovalRequest> {
  return command<ApprovalRequest>('get_approval_request', { requestId });
}
export async function approveRequest(
  requestId: string,
  reviewerId: string,
  reason?: string,
): Promise<void> {
  return command<void>('approve_request', { requestId, reviewerId, reason });
}
export async function rejectRequest(
  requestId: string,
  reviewerId: string,
  reason: string,
): Promise<void> {
  return command<void>('reject_request', { requestId, reviewerId, reason });
}
export async function requiresApproval(action: ApprovalAction): Promise<boolean> {
  return command<boolean>('requires_approval', { action });
}
export async function calculateRiskLevel(action: ApprovalAction): Promise<string> {
  return command<string>('calculate_risk_level', { action });
}
export async function getApprovalStatistics(teamId?: string): Promise<ApprovalStatistics> {
  return command<ApprovalStatistics>('get_approval_statistics', { teamId });
}
export async function expireTimedOutRequests(): Promise<number> {
  return command<number>('expire_timed_out_requests');
}

// ---- Privacy ----

export async function settingsUpdatePrivacy(preferences: PrivacyPreferences): Promise<void> {
  return command<void>('settings_update_privacy', { preferences });
}
export async function privacyExportData(): Promise<string> {
  return command<string>('privacy_export_data');
}
export async function privacyDeleteAccount(userId: string): Promise<string> {
  return command<string>('privacy_delete_account', { userId });
}
